[RFC] Implementation of dense neural network as CV

HanatoK opened the pull request:
This pull request introduces a new CV `NeuralNetwork` to compute the output of DNN. The weights and biases of the DNN can be specified by text files as documented in the modified manual.

I add `[RFC]` in the title because I also implemented a general sub-CV, `customColvar`, to support various normalization methods  by lepton that cannot be done by `linearCombination`, and I don't know if this is the right way to do so.

jhenin commented:
Since Lepton is not coming to VMD anytime soon, and there is a small number of very common nonlinearity functions, I think we should consider hard-coding the most useful ones and have this working as often as possible without Lepton.

giacomofiorin commented:
FYI it would be possible to add Lepton using the patch here:
https://github.com/giacomofiorin/vmd-patches
but I don't believe that change will make into the official VMD source tree and builds anytime soon.

HanatoK commented:
Hi @jhenin and @giacomofiorin , I made a fork at https://github.com/HanatoK/lepton_nocpp11 that should not require C++11. I have just tested with `-std=c++98` and it works, but I have not tested all the functionalities.

giacomofiorin commented:
Try opening another PR, which you may close later if you like. NB: C++11 would be okay, there is already significant code from you that is inside `#ifdef` anyway. 

It's Lepton that is the bigger issue, because it can't be integrated with VMD's build system as-is (unlike for NAMD, LAMMPS, etc)

HanatoK commented:
> Try opening another PR, which you may close later if you like. NB: C++11 would be okay, there is already significant code from you that is inside `#ifdef` anyway.
> 
> It's Lepton that is the bigger issue, because it can't be integrated with VMD's build system as-is (unlike for NAMD, LAMMPS, etc)

Lepton is not a part of Colvars, so I cannot make a PR for that. Maybe I can ask John to integrate the lepton code without C++11, but that's another problem. As for this PR, I think that it is fine even the new CV is only available in NAMD for now.

jhenin commented:
You can always talk to John about Lepton, but I'm not hoping for a quick result. My question is: is the flexibility of Lepton essential for this feature, or would hard-coded functions work?

HanatoK commented:
@jhenin Hard-coded function, all existing CVs and linear combination work.

jhenin commented:
> Maybe I can ask John to integrate the lepton code without C++11

To be clear, in the near future, John will be enabling C++11 in his builds, so the C++11 problem is going away, but Lepton is going to be a much more complicated story.

HanatoK commented:
> > Maybe I can ask John to integrate the lepton code without C++11
> 
> To be clear, in the near future, John will be enabling C++11 in his builds, so the C++11 problem is going away, but Lepton is going to be a much more complicated story.

Never mind. I have used `#ifdef LEPTON` in the code, and I think it should be compiled without lepton, although the functionalities are somewhat limited.

jhenin commented:
VMD doesn't build without Lepton:
```
colvar_neuralnetworkcompute.C: In member function ‘void neuralnetworkCV::denseLayer::compute(const std::vector<double>&, std::vector<double>&) const’:
colvar_neuralnetworkcompute.C:156:25: error: ‘m_custom_activation_function’ was not declared in this scope
             output[i] = m_custom_activation_function.evaluate(output[i]);
```

HanatoK commented:
> VMD doesn't build without Lepton:
> 
> ```
> colvar_neuralnetworkcompute.C: In member function ‘void neuralnetworkCV::denseLayer::compute(const std::vector<double>&, std::vector<double>&) const’:
> colvar_neuralnetworkcompute.C:156:25: error: ‘m_custom_activation_function’ was not declared in this scope
>              output[i] = m_custom_activation_function.evaluate(output[i]);
> ```
> 
> but with Lepton it builds fine.

Sorry, I missed some ifdef macros after adding support of custom activation functions. It should be solved by the latest commit.

giacomofiorin commented on src/colvar.cpp:
```diff
@@ -886,7 +886,9 @@ int colvar::init_components(std::string const &conf)
   error_code |= init_components_type<euler_phi>(conf, "euler phi angle of the optimal orientation", "eulerPhi");
   error_code |= init_components_type<euler_psi>(conf, "euler psi angle of the optimal orientation", "eulerPsi");
   error_code |= init_components_type<euler_theta>(conf, "euler theta angle of the optimal orientation", "eulerTheta");
+  #if (__cplusplus >= 201103L)
```
Shouldn't this also be wrapped in `#ifdef LEPTON`?

HanatoK replied:
I added `#ifdef LEPTON` in the latest commit, although the code can be compiled without it and fallback to `linearCombination`.

giacomofiorin replied:
Good, thanks. We're in a fairly good place now, John says that we can rely on C++11 for new features, as he'll make at least one C++11-enabled VMD build for Linux that people can use. We basically need to make sure that VMD as a whole doesn't break for people on very old machines (past EOL), which by now ought to accept that they'll get reduced functionality. Most VMD users will get the full set of features.

A Lepton dependency remains trickier, as that *will* prevent distribution with VMD.

HanatoK reviewed (commented):
Thanks for adding these new functions. I hesitated to add them because their derivatives are not mathematically continuous.

jhenin commented:
I am seeing double-free problems when deleting these variables. What I added in the lastest commit solved some problems but not all of them. I still see this warning: "Warning: destroying "atom group group1" before its parents objects:", and segfaults. There is a related issue with dependencies.
Can you make sure that there is a deletion path that works correctly? This is really testing the usual hierarchy colvar->cvcs. Maybe at some point we get tired of patching this individually, and address it in a general way, but I can't see how.

HanatoK commented:
@jhenin I also encounter a crash even in the initialization. I must miss something. Maybe I should not use the following code:
```
template<typename def_class_name> int colvar::init_components_type(std::string const &conf,
                                                                   char const * /* def_desc */,
                                                                   char const *def_config_key)
{
#if (__cplusplus >= 201103L)
  global_cvc_map[def_config_key] = [](const std::string& cvc_conf){return new def_class_name(cvc_conf);};
#endif
```
This means if a CV is `linearCombination` of `azpathCV`, `customColvar` or some types after `linearCombination` in the following lines, the code doesn't work and crashes.
```
  error_code |= init_components_type<linearCombination>(conf, "linear combination of other collective variables", "linearCombination");
  error_code |= init_components_type<gspathCV>(conf, "geometrical path collective variables (s) for other CVs", "gspathCV");
  error_code |= init_components_type<gzpathCV>(conf, "geometrical path collective variables (z) for other CVs", "gzpathCV");
  error_code |= init_components_type<aspathCV>(conf, "arithmetic path collective variables (s) for other CVs", "aspathCV");
  error_code |= init_components_type<azpathCV>(conf, "arithmetic path collective variables (s) for other CVs", "azpathCV");
  error_code |= init_components_type<euler_phi>(conf, "euler phi angle of the optimal orientation", "eulerPhi");
  error_code |= init_components_type<euler_psi>(conf, "euler psi angle of the optimal orientation", "eulerPsi");
  error_code |= init_components_type<euler_theta>(conf, "euler theta angle of the optimal orientation", "eulerTheta");
  #if (__cplusplus >= 201103L)
  #ifdef LEPTON
  error_code |= init_components_type<customColvar>(conf, "CV with support of the lepton custom function", "customColvar");
  #endif
```

HanatoK commented:
I will take some time to look into the problem.

HanatoK commented:
@giacomofiorin I split the `init_components_type` into two code paths in the commit https://github.com/Colvars/colvars/pull/444/commits/027d843ce1c43e8f32a6d07479d75a2501c0e6a1. Is that feasible?
@jhenin Could you provide more information about how to reproduce the crash after your commit https://github.com/Colvars/colvars/pull/444/commits/30db79019ae27e839d91b5280f3d894b58f084dd?

giacomofiorin commented:
No, please let's not duplicate a big function with only one line different *and* that line is conditionally compiled anyway. Plus the `#ifdef`s may as well go out the window in the near future. We're basically just waiting for the VMD release, for which John says he will make with both C++11 enabled and C++11 disabled. I really don't care about the latter: but as soon as a functional VMD official build is released, I would gladly kill all the `#ifdef`s for the future.

If I understand correctly, you want the global map of CVC names to be fully populated when you are initialing the CVCs for the first colvar. That makes sense, and it'd be overall helpful to have the map exposed through the scripting interface. I'll take a stab at it, since it's something that I really would like to have.

jhenin commented:
In short: now we can (finally) forget about the C++11 issue. The current ifdefs (before the latest commit) are sufficient, and as long as a legacy version of VMD can compile, we don't mind if it is missing an arbitrary number of features.

I will provide you with a test case for the double-delete situation.

HanatoK commented:
> In short: now we can (finally) forget about the C++11 issue. The current ifdefs (before the latest commit) are sufficient, and as long as a legacy version of VMD can compile, we don't mind if it is missing an arbitrary number of features.
> 
> I will provide you with a test case for the double-delete situation.

But the problem is that without the latest commit (https://github.com/Colvars/colvars/pull/444/commits/027d843ce1c43e8f32a6d07479d75a2501c0e6a1), the code doesn't work if a sub-CVC is not in the map if the initialization of parent CVC happens first. For example:
```
    linearCombination {
        name xxx

        customColvar {
            name 001
            customFunction sin(phi_1/180.0*3.1415926535897)
            dihedral {
                name phi_1
                group1 {atomNumbers {5}}
                group2 {atomNumbers {7}}
                group3 {atomNumbers {9}}
                group4 {atomNumbers {15}}
                scalable off
            }
        }
        customColvar {
            name 002
            customFunction sin(phi_2/180.0*3.1415926535897)
            dihedral {
                name phi_2
                group1 {atomNumbers {15}}
                group2 {atomNumbers {17}}
                group3 {atomNumbers {19}}
                group4 {atomNumbers {25}}
                scalable off
            }
        }
    }
```

giacomofiorin commented:
> But the problem is that without the latest commit ([027d843](https://github.com/Colvars/colvars/commit/027d843ce1c43e8f32a6d07479d75a2501c0e6a1)), the code doesn't work if a sub-CVC is not in the map if the initialization of parent CVC happens first. For example:

Correct, which is why I'm working toward initializing the full CVC map before any CVC object is constructed, but without duplication and also with the ability to expose the CVC map during parsing and scripting. If we're going through the hassle of editing a core piece of the code, it better give multiple benefits beyond this PR.

HanatoK commented:
> No, please let's not duplicate a big function with only one line different _and_ that line is conditionally compiled anyway. Plus the `#ifdef`s may as well go out the window in the near future. We're basically just waiting for the VMD release, for which John says he will make with both C++11 enabled and C++11 disabled. I really don't care about the latter: but as soon as a functional VMD official build is released, I would gladly kill all the `#ifdef`s for the future.
> 
> If I understand correctly, you want the global map of CVC names to be fully populated when you are initialing the CVCs for the first colvar. That makes sense, and it'd be overall helpful to have the map exposed through the scripting interface. I'll take a stab at it, since it's something that I really would like to have.

Please have a look at my latest commit. I remove duplicate code by using some awkward `#ifdef`s.

giacomofiorin commented:
Well, I was already working on that, and `916ac01` now created conflicts for me, and now I can't get that done as quickly, sorry. If your quick fix works for the issue of the map being incomplete, I can just stash my work away and resolve those conflicts later. There is still the issue of the double-free to look at anyway.

HanatoK commented:
> I will provide you with a test case for the double-delete situation.

@jhenin Do you upload the test case somewhere on github or send it to me by email?

HanatoK commented:
@jhenin Not sure if the latest commit solves your problem. In my tests there is no warning in deletion.

jhenin commented:
Thank you @HanatoK , I don't see the deletion issues in my tests anymore. Good idea of you to document the code structure better!

HanatoK commented:
> Can you add return statements to prevent the constructor from proceeding after an error?

Done.

giacomofiorin commented on namd/tests/library/030_neuralnetwork/test.in:
```diff
@@ -0,0 +1,107 @@
+colvarsTrajFrequency    1
```
This configuration is quite portable, but it is run by a job script to NAMD (2.x branch) and does not test a restarting job. This means that in the current setup it won't be tested for other backends.

Can you make an input based on decaalanine, so that we can test the other backends as well? I can propagate it if you like but would need updated .txt files.

HanatoK replied:
I change the input for testing to deca-alanine.

giacomofiorin reviewed (changes requested):
I'm generally okay with this, save for the small issues. Disclaimer: not being familiar with the methodology, I only commented about general implementation issues.

Note that I moved the citation to `colvars-code-refs.bib`, where references to code implementation are located, which will be also printed at runtime.

Some of the build recipes are still not up to date:
```
make: *** No rule to make target `colvars/src/colvarcomp_neuralnetwork.h', needed by `obj/colvarcomp_neuralnetwork.o'.  Stop.
```

HanatoK commented on src/colvarcomp.h:
```diff
@@ -28,6 +28,7 @@
 #if (__cplusplus >= 201103L)
 // C++11-only functions
 #include "colvar_geometricpath.h"
+#include "colvar_neuralnetworkcompute.h"
```
Done. I use `std::unique_ptr` which should be allowed in C++11.

HanatoK commented on src/colvarcomp_neuralnetwork.cpp:
```diff
@@ -0,0 +1,375 @@
+
+using namespace neuralnetworkCV;
+
+colvar::customColvar::customColvar(std::string const &conf): linearCombination(conf) {
```
I think both `customColvar` in `colvarcomp_neuralnetwork.cpp` and `linearCombination` in `colvarcomp_gpath.cpp` should be in a separate file, but don't know which file they should be in. Could you propose an existing one or a new filename for them?

HanatoK replied:
As far as I know muparser does not have automatic differentiation for a given math expression, which should be required for CV calculations. How do you plan to use muparser in Colvars?

jhenin replied:
I think `linearCombination` and `customColvar` could go to files colvarcomp_combination.{h,cpp}.

HanatoK commented on src/colvarcomp_neuralnetwork.cpp:
```diff
@@ -0,0 +1,375 @@
+    size_t l = 0;
+    for (size_t i = 0; i < x.size(); ++i) {
+        for (size_t i_cv = 0; i_cv < cv.size(); ++i_cv) {
+            colvarvalue current_cv_value(cv[i_cv]->value());
```
Done. Their values are unchanged so I use `const colvarvalue&`.

HanatoK commented on src/colvarcomp_neuralnetwork.cpp:
```diff
@@ -0,0 +1,375 @@
+#include <cmath>
+#include <cstdlib>
+#include <limits>
+#include <utility>
```
Done.

HanatoK commented:
@jhenin @giacomofiorin I have moved `linearCombination` and `customColvar` to a new file. As for the commit message I made a mistake so I corrected it by `git commit --amend` and `git push --force`, but it introduced a new conflict, which I solved later. Now the branch looks fine but there's an extra commit https://github.com/Colvars/colvars/pull/444/commits/427a9f787641a7e90d411d3f274b199a744b037d

giacomofiorin commented:
I'm fine with the merge commit. From the look of it it wouldn't take too much work to rebase it, but it's not a high priority for me. No warnings on my offline tests aside from one with Clang 13, which is unrelated to this PR anyway (I'll push a fix separately).

Did you two figure out why there were double free warnings?

If all is good please merge as-is.

giacomofiorin commented:
Actually, I forgot the previous comments, so yeah, let's merge. Actually the rebase wasn't too bad. I put the result here:
https://github.com/Colvars/colvars/tree/neural_network-rebase
If it looks right to you (it does to me) we can merge this PR the same way.

jhenin commented:
Your merged branch looked good so I just grabbed it @giacomofiorin  :-)

HanatoK commented:
Thanks for merging! However now I notice there's a small conflict between this PR and #447 . Both PRs make their NAMD tests starting from `030_*` (you can view them in https://github.com/Colvars/colvars/tree/master/namd/tests/library). Is that a problem?

jhenin commented:
Good catch!
